crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Generic.hs

```diff
@@ -1292,6 +1297,47 @@ mergeMem c x y =
       in x & memState .~ s'
     _ -> error "mergeMem given unexpected memories"
 
+--------------------------------------------------------------------------------
+-- Finding allocations
+
+-- When we have a concrete allocation number, we can ask more specific questions
+-- to the solver and get (overapproximate) concrete answers.
+
+data SomeAlloc sym =
+  forall w. SomeAlloc AllocType Natural (Maybe (SymBV sym w)) Mutability Alignment String
+
+ppSomeAlloc :: forall sym. IsExprBuilder sym => SomeAlloc sym -> Doc
+ppSomeAlloc (SomeAlloc atp base sz mut alignment loc) =
+  ppAlloc (Alloc atp base sz mut alignment loc :: MemAlloc sym)
+
+-- | Find an overapproximation of the set of allocations with this number.
+--
+--   Ultimately, only one of these could have happened.
+--
+--   It may be possible to be more precise than this function currently is.
+--   In particular, if we find an 'Alloc' with a matching block number before a
+--   'AllocMerge', then we can (maybe?) just return that 'Alloc'. And if one
+--   call of @helper@ on a 'MemAlloc' returns anything nonempty, that can just
+--   be returned (?).
+possibleAllocs ::
+  forall sym .
+  (IsSymInterface sym) =>
+  Natural              ->
+  Mem sym              ->
+  [SomeAlloc sym]
+possibleAllocs n =
+  let helper m =
```

Comment:
Why bind `m` here? It looks like it is always equal to `n`...

Reply:
Also, I usually prefer to use a `where` binding for this sort of pattern rather than a `let`

Reply:
@robdockins I was also confounded by the need to bind `m`. Here's what happens when I get rid of the binding an parameter:
```
src/Lang/Crucible/LLVM/MemModel/Generic.hs:(1330,9)-(1339,64): error:
    * Couldn't match type `t0 (MemAlloc sym0)' with `Natural'
      Expected type: Natural -> [MemAlloc sym0] -> [SomeAlloc sym1]
        Actual type: t0 (MemAlloc sym0) -> [SomeAlloc sym1]
    * In the expression:
        foldMap
          $ \case
              MemFree _ -> []
              Alloc atp base sz mut alignment loc
                -> if base == n then
                       [SomeAlloc atp base sz mut alignment loc]
                   else
                       []
              AllocMerge (asConstantPred -> Just True) as1 as2 -> helper n as1
              AllocMerge (asConstantPred -> Just False) as1 as2 -> helper n as2
              AllocMerge _ as1 as2 -> helper n as1 ++ helper n as2
```